server/src/handlers/prom.rs

```diff
@@ -0,0 +1,427 @@
+    type Context = RequestContext;
+    type Err = Error;
+
+    /// Write samples to remote storage
+    async fn write(&self, ctx: Self::Context, req: WriteRequest) -> Result<()> {
+        let request_id = RequestId::next_id();
+        let deadline = ctx.timeout.map(|t| Instant::now() + t);
+        let catalog = &ctx.catalog;
+        self.instance.catalog_manager.default_catalog_name();
+        let schema = &ctx.schema;
+        let schema_config = self
+            .schema_config_provider
+            .schema_config(schema)
+            .context(SchemaError)?;
+        let plans = crate::grpc::storage_service::write::write_request_to_insert_plan(
+            request_id,
+            catalog,
+            schema,
+            self.instance.clone(),
+            Self::convert_write_request(req)?,
+            schema_config,
+            deadline,
+        )
+        .await
+        .context(GRPCWriteError)?;
+
+        let mut success = 0;
+        for insert_plan in plans {
+            success += crate::grpc::storage_service::write::execute_plan(
+                request_id,
+                catalog,
+                schema,
+                self.instance.clone(),
+                insert_plan,
+                deadline,
+            )
+            .await
+            .context(GRPCWriteError)?;
+        }
+        debug!(
+            "Remote write finished, catalog:{}, schema:{}, success:{}",
+            catalog, schema, success
+        );
+
+        Ok(())
+    }
+
+    /// Process one query within ReadRequest.
+    async fn process_query(&self, ctx: &Self::Context, q: Query) -> Result<QueryResult> {
+        let mut filters = Vec::with_capacity(q.matchers.len());
+        filters.push(format!(
+            "{} between {} AND {}",
+            TIMESTAMP_COLUMN, q.start_timestamp_ms, q.end_timestamp_ms
+        ));
+        let mut measurement = None;
+        for m in &q.matchers {
+            if m.name == NAME_LABEL {
+                measurement = Some(m.value.to_string());
+                continue;
+            }
+
+            let filter = match m.r#type() {
+                label_matcher::Type::Eq => format!("{} = '{}'", m.name, m.value),
+                label_matcher::Type::Neq => format!("{} != '{}'", m.name, m.value),
+                // https://github.com/prometheus/prometheus/blob/2ce94ac19673a3f7faf164e9e078a79d4d52b767/model/labels/regexp.go#L29
+                label_matcher::Type::Re => format!("{} ~ '^(?:{})'", m.name, m.value),
+                label_matcher::Type::Nre => format!("{} !~ '^(?:{})'", m.name, m.value),
+            };
+            filters.push(filter)
+        }
+
+        let measurement = measurement.context(MissingName).unwrap();
+        let sql = format!(
+            "select * from {} where {} order by {}, {}",
+            measurement,
+            filters.join(" and "),
+            TSID_COLUMN,
+            TIMESTAMP_COLUMN
+        );
```

Comment:
Actually, I guess a separate method can be defined for this conversion and afterwards we can add some unit tests for it if you will.